### ISSUE TITLE
v0.3.22: reconstruct/promotion deadlock traps serve-http in a permanent restart loop — promotion guard false-positives because reconstruction renumbers agent IDs

### ISSUE THREAD

**Issue by jmcmacnz** [Dicklesworthstone/mcp_agent_mail_rust#208]
> **Authorship note:** this report was researched and written by an AI coding agent (Claude) working on the affected host, and reviewed by me before posting. Every log excerpt, table, and SQL result below is verbatim output from real runs on that machine rather than reconstructed from memory. If any inference looks wrong, say so and I can re-run the measurement.

## Summary

After a host reboot, the corruption-recovery path deadlocked against itself and the service could not be started again without manual SQLite surgery. `serve-http` restarted forever under systemd, redoing a full archive reconstruction on every iteration before failing.

Two startup guards each demand the opposite thing:

- **Gate A — reconstruct trigger.** The live DB is judged incomplete relative to the Git archive, so a reconstruction is required. In effect it wants `live ⊇ archive`.
- **Gate B — promotion guard.** The freshly reconstructed candidate is judged to lose "stable coordination keys" that exist in the live DB, so promotion is refused and the process exits 1. In effect it wants `candidate ⊇ live`.

When the two stores have each drifted in a different direction, neither condition can be satisfied: A insists on rebuilding, B refuses to install the rebuild, and there is no third path. The process never reaches readiness, so none of the normal self-healing (backup rotation, search backfill, doctor auto-fix) ever runs.

The practical severity is that this is **unrecoverable rather than degraded** — there is no read-only or safe-mode fallback, and both `am doctor repair` and `am doctor reconstruct` route back into the same refusal.

## Environment

- Official release asset `v0.3.22`, `x86_64-unknown-linux-gnu`, kernel 6.17.0-41-generic, ext4
- systemd user unit, `Restart=always`, `RestartSec=5`
- Mailbox: 894 messages, 105 agents, 4 projects, 1203 recipients, 956 file reservations
- Triggered by an ordinary host reboot

I only observed this on v0.3.22; I cannot say whether earlier versions can reach the same state, because by the time I rolled back to v0.3.21 the DB had already been repaired by hand.

For context (not part of this report): the index corruption that *triggers* the recovery path appears to recur on this host on every reboot. In the journal's full retained history, all four boots produced a corruption or quarantine artifact within the same minute as boot — 06-17, 06-23, 07-03 and 07-25. Previous occurrences self-healed because reconstruction succeeded; this time the promotion guard refused it, which is what turned a routine recovery into an outage.

## Logs

Initial form — corruption detected, rebuild refused:

```
Startup detected mailbox database issues; reconstructing from archive
  (Health probes failed for .../storage.sqlite3 (possible corruption);
   archive recovery is available under .../projects)
database reconstruction from archive complete stats=reconstructed 4 projects,
  101 agents, 894 messages (1203 recipients), 0 parse errors;
  skipped 9 duplicate canonical file(s) across 9 message id(s)
error: reconstruction succeeded but final database promotion failed:
  recovery candidate .../storage.sqlite3.reconstruct-20260725_012642_207.sqlite3
  would lose stable coordination keys from .../storage.sqlite3
  (projects=0, products=0, product_project_links=0, agents=39, contacts=45,
   reservations=12, messages=66, message_recipients=721,
   proof_gate_consumed_nonces=0); refusing promotion
agent-mail.service: Main process exited, code=exited, status=1/FAILURE
agent-mail.service: Scheduled restart job, restart counter is at 32.
```

The reported corruption was index-only — every line of `PRAGMA integrity_check` was of the form `wrong # of entries in index …` on `messages` / `message_recipients`, with no table or page damage. A plain `REINDEX` cleared it completely (`integrity_check` → `ok`).

After `REINDEX`, Gate A simply switched reasons and the loop continued:

```
Startup detected mailbox database issues; reconstructing from archive
  (Archive canonical data is not fully represented in the SQLite DB
   (messages archive=894 db=891); reconstruct from archive .../projects)
error: reconstruction succeeded but final database promotion failed: … would lose
  stable coordination keys … (reservations=12, messages=76,
  message_recipients=306, …); refusing promotion
```

So Gate A wanted `db ⊇ archive` (894 vs 891) while Gate B wanted `candidate ⊇ db`.

## The promotion guard's verdict looks like a false positive

This is the part I think is most worth your attention. I `ATTACH`ed the rejected candidate alongside the live DB and compared the two directly:

```sql
canon_msgs                   891
cand_msgs                    894
same_id_same_subject_and_ts  891
same_id_DIFFERENT_content      0
canon_only_by_id               0   -- live DB has NOTHING the candidate lacks
cand_only_by_id                3
```

The candidate was a **strict superset** of the live DB for messages — every one of the 891 live messages was present in the candidate under an identical `id`, `subject` and `created_ts`, and the candidate held 3 more. Promotion would have lost zero messages. Yet the guard claimed it would lose 76 messages and 306 recipients.

The likely reason: **reconstruction renumbers agent IDs**, so any comparison keyed on agent identity mis-joins.

```sql
agents_id_name_match         1
agents_id_name_MISMATCH    100   -- of 101 shared ids
```

Concretely, live `agents.id=39` was `SunnyShore` while candidate `id=39` was `CyanBay`; `id=81` was `BlackFern` vs `PearlBirch`. Since `message_recipients` is keyed `(message_id, agent_id)`, a naive comparison across the two DBs sees almost every recipient row as "missing", which matches the inflated `message_recipients=721` / `306` figures in the refusal. This smells related to the ID-reassignment behaviour fixed for projects in #187.

If that is right, the guard is protecting against data loss that would not occur, and it does so in a way that is unrecoverable rather than degraded.

## Blast radius

`Restart=always` + a deterministic startup failure means the reconstruction runs on every restart. The restart counter reached **35 in about 4 minutes**, leaving ~48 abandoned `storage.sqlite3.reconstruct-*.sqlite3` candidates (~1.7 MB each) plus a quarantined activity-lock file per iteration.

On this host the same storage root also holds artifacts from earlier recovery events at 5.5 GB (`storage.sqlite3.corrupt-20260615…`) and 6.1 GB (`storage.sqlite3.reconstruct-failed-20260617…`), which had pushed the filesystem to 83% full. Startup backup-rotation reclaimed 6.1 GB by itself once the service finally came up (`rotated storage backups removed=45 kept=12 bytes_reclaimed=6177132544`), so the cleanup logic works — it just never gets to run while the process is dying before readiness.

## What actually recovered it

Neither `am doctor repair` nor `am doctor reconstruct` could break the loop: `repair` gates on `quick_check_ok` and, since that fails, falls straight back into the same archive reconstruction that Gate B then refuses.

Manual sequence that worked:

1. `systemctl --user stop agent-mail` (mutating `doctor` ops need the exclusive activity lock).
2. `sqlite3 storage.sqlite3 "REINDEX;"` — clears index-only damage losslessly; `integrity_check` → `ok`.
3. Copy the 3 archive-only messages from the rejected candidate into the live DB **remapping `sender_id`/`agent_id` through agent names**, since IDs are not comparable across the two DBs. (`recipients_json` stores names, so it is safe verbatim.)
4. `integrity_check` + `foreign_key_check` clean → service starts and stays up.

## Suggestions

1. Make the promotion guard's "would lose" comparison identity-stable — resolve agents by name (or a stable external key) rather than row ID, since reconstruction renumbers them.
2. Break the A/B deadlock explicitly: if the candidate is a superset by the stable key, promote; if the two genuinely diverge, fall back to a **merge** (or come up degraded/read-only) instead of exiting 1 forever.
3. Consider `REINDEX` as a first-line repair when `integrity_check` reports only index-entry-count mismatches — it is lossless and avoids the whole reconstruction path.
4. Consider a startup failure backstop (systemd `StartLimitBurst`, or an internal "N consecutive identical startup failures → stop retrying") so a deterministic failure cannot spin. I added `StartLimitIntervalSec=300` / `StartLimitBurst=5` locally as a stopgap.

**Comment by Dicklesworthstone** [comment]
Confirmed the false-positive, with a refinement to the mechanism (from reading `forensics.rs`'s continuity sets — `collect_recovery_continuity_sets` / `compare` path):

The guard's comparison is **not** actually keyed on raw agent row IDs — the snapshot queries already resolve identity through names (`JOIN agents recipient ON recipient.id = mr.agent_id` then fingerprint `recipient.name`), and message fingerprints deliberately exclude numeric ids for remap-invariance. What *does* diverge across a reconstruction is the **semantic payload inside the fingerprints**:

- the agents fingerprint includes `registration_token_sha256` — reconstruction can't recover the live token (and registration rotates it), so every agent whose token drifted counts as "lost" (your 39/101);
- message fingerprints include `body_md`, `recipients_json` and `attachments` verbatim — any canonical-form normalization during archive replay makes the row count as "lost" (your 76);
- recipient fingerprints chain to their **parent message's semantic digest** and include `read_ts`/`ack_ts`, which the archive does not persist — so one drifting message field amplifies into every one of its recipient rows "lost" (your 721/306), and read-state loss is conflated with row loss.

So the effect you measured is right (candidate was a strict superset by stable key; the refusal was protecting against loss that would not occur), but the fix needs to split each category's fingerprint into an **identity key** (project slug + agent name; message stable key; recipient = message identity + recipient name + kind) and a **volatile payload** (tokens, read/ack timestamps, serialized bodies) — refuse promotion only on identity loss, and report volatile drift as informational. That also addresses your suggestion 2's "superset by stable key ⇒ promote".

Your suggestions 3 (REINDEX as first-line repair for index-only corruption) and 4 (startup failure backstop) stand on their own and are tracked separately.

Tracked as br-87tol; the unrelated read-path outage that put this host on v0.3.22 is fixed in v0.3.24, which does not change this guard's behavior.

**Comment by Dicklesworthstone** [comment]
Implemented — all three legs of this are now addressed on `main` (96f82b15):

**1. Identity/volatile fingerprint split (the false-positive refusal).** `RecoveryContinuitySets` now carries identity-key projections alongside the full payload fingerprints, and the promotion guard refuses **only on identity loss**:

- agents: `project slug/human_key + agent name` (registration-token hash and reaper flag are volatile — the archive can't recover a live token, which was your 39/101);
- messages: `project + sender + thread_id + subject + created_ts`, hashed (body/importance/ack flag/serialized recipients/attachments are volatile — canonical-form normalization was your 76);
- recipients: chained to the parent message's **identity** digest + recipient name + kind (read/ack timestamps volatile — this stops one drifted message from amplifying into hundreds of "lost" recipient rows, your 721/306);
- reservations: `project + agent + path_pattern + exclusive + created_ts` (reason/renewal/release state volatile);
- contacts: the two endpoints (status/reason/timestamps volatile);
- projects, products, product-project links, and proof-gate nonces keep the strict full-payload contract.

Volatile drift is still visible — the refusal message and the recovery receipt now carry it as informational counts (`identity_*` fields; old receipts round-trip byte-identically so existing receipt chains keep verifying), and a promotion through drift logs a warning. Your "superset by stable key ⇒ promote" is exactly what the guard now does; the duplicate-replay check also moved to identity keys so payload normalization can't disguise a real replay.

The regression test reproduces your exact shape — renumbered agent ids, nulled tokens, dropped read state, normalized body, plus archive-only rows — and asserts the candidate promotes with a finalized, verifying receipt (`recovery_receipt_promotes_identity_superset_despite_volatile_payload_drift`).

**2. REINDEX as first-line repair.** This existed since v0.2.10 (`try_repair_index_only_corruption`) but was gated on `PRAGMA quick_check` details — and `quick_check` *skips index-content-vs-table verification*, which is precisely the class you hit ("wrong # of entries in index"). So the health probes (which run full `integrity_check`) declared the file corrupt while the repair probe saw "ok" and declined, and every boot escalated to full reconstruction. The gate now consults the full `integrity_check` details when `quick_check` is clean; an integration test creates real index/table divergence by flipping table-page bytes under a live index and proves the layered probes flag it, REINDEX heals it, and no reconstruction runs.

**3. Startup failure backstop.** Already shipped: the unit `am service install` generates carries `StartLimitBurst=5` / `StartLimitIntervalSec=300` / `RestartSec=30` (added for #96). Your unit was hand-written with `Restart=always` + `RestartSec=5`, which is why the spin was unbounded — your local `StartLimitIntervalSec=300`/`StartLimitBurst=5` stopgap matches what the generated unit does.

With 1 + 2 together your host's sequence becomes: index-only corruption at boot → in-place REINDEX (no reconstruction), and if a reconstruct is ever genuinely needed, an identity-superset candidate promotes instead of looping. `docs/RECOVERY_RUNBOOK.md` now documents the refusal-loop symptom and the manual `REINDEX` remediation for v0.3.22–v0.3.24.

Ships in the next release (v0.3.25). Thanks for an exceptional report — the ATTACH-comparison table made the false positive undeniable, and the "guard is protecting against loss that would not occur, in a way that is unrecoverable rather than degraded" framing is what drove the fix design.

**Comment by Dicklesworthstone** [comment]
This was fixed on main shortly after your report — closing with the specifics so the mechanism is on record.

**Root cause (matches your Gate A/Gate B analysis):** reconstruction renumbers agent IDs, and the promotion guard treated ID-keyed "stable coordination keys" present in the live DB but absent (under the same IDs) from the candidate as loss — refusing promotion of a candidate that actually contained the same identities under new IDs. With Gate A (reconstruct trigger: live ⊇ archive) and Gate B (promotion guard: candidate ⊇ live) each demanding the opposite inclusion after divergent drift, `serve-http` looped forever under `Restart=always`, redoing a full reconstruction per iteration with no read-only fallback.

**Fixes on main:**
- 96f82b15 — the promotion guard now refuses **only on genuine identity loss** (stable identity keys, not renumbered rowids), so a reconstruction that renumbers agents no longer false-positives; the REINDEX gate consults full `integrity_check` rather than `quick_check` (the index/table-desync class that *triggers* this recovery typically passes quick_check).
- e5cc9539 — archive-identity drift healing + failed candidates are preserved (not deleted) for diagnosis instead of being rebuilt from scratch each restart.
- Additionally, startup restart-coordination now fails closed when another live process holds the storage-root lock mid-boot instead of proceeding into the kill path (the #253 outage class), so supervisors retry cleanly rather than racing a reconstructing peer.

End-to-end regression tests exist for the renumbering scenario (`crates/mcp-agent-mail-db/src/pool.rs`, GH#208 repro block: real index-vs-table divergence, reconstruct, promotion succeeds).

This ships in the next release after v0.3.24. If a two-gate deadlock reappears on a fixed build, please reopen with the two gate log lines — the remaining theoretical case (true identity loss in both directions) intentionally still refuses, but it should name exactly which identities it believes are at risk.
